documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,183 @@
+# Event Pipe
+
+## Overview
+
+> TODO: Write an overview of what this does
+
+## TODOs
+* formalize struct names
+* formalize error codes (Use HRESULT values?)
+
+## Inter-Process Communication Specification
+
+### Current
+
+```c++
+enum class DiagnosticMessageType : uint32_t
+{
+    // EventPipe
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+
+### Proposed
+
+Connection:
+```
+runtime <- client : handshake (set length byte array that represents version)
+runtime -> client : sends handshake back as ACK
+runtime <- client : packet containing start code
+runtime -> client : ACK + potential payload
+repeat till closed
+```
+
+### Transport considerations
+
+* All data will be sent in the endianess of the machine
+* All data will be sent in a packet containing a header of information concerning the contents
```

Comment:
I'm not certain what this means? Given how we did event pipe trace streaming it may not be true ; )

Reply:
These points are more "wants" than anything.  Since the client will presumably be on the same machine as the runtime, it makes things easier to keep things in the same endianness as the machine.  The argument could be made that it makes more sense to arbitrarily make everything little-endian, though.  In this case, there will be no question about the endianness and it will be up to the consumer to encode/decode that detail.

As for the point on packets, that is certainly not the case right now.  It shouldn't be a big change to what is happening for something like event pipe though.  Pretty much every packet for event pipe would end up looking the same, `[header][payload ..... ]` where the payload is just what is being sent now.  The event pipe encoding is a higher layer protocol/encoding and sending it via this IPC protocol shouldn't affect its use.